text/0010-trits/0010-trits.md

```diff
@@ -0,0 +1,124 @@
+
+To allow implementing all conversions without having to focus too much on unnecessary work, we propose the following requirements:
+- a conversion from any encoding to the same encoding is done by copy;
+- a conversion from `t1b1` to any other encoding is efficiently implemented by `From` trait;
+- a conversion from any encoding to `t1b1` is efficiently implemented by `From` trait;
+- a conversion, proven to be heavily used, from any encoding to any other encoding is efficiently implemented by `From` trait with specialization support;
+- any other conversion falls back to a trivial conversion with an intermediary `t1b1` step;
+
+Note that inner structure should not provide any method to public. It should be
+implemented by methods of `Trits<T>` to that specific type. This interface can
+dereference to inner structure and provide more functionality other than general
+methods.
+
+```rust
+pub trait Encoding {
+    fn with_capacity(n: usize) -> Self;
+    fn len(&self) -> usize;
+    ...
+}
+
+pub struct T1B1 {
+    raw: ...,
+}
+
+impl Encoding for T1B1 {
+    ...
+}
+
+pub struct Trits<T: Encoding>{
+    inner: T,
+}
+
+impl<T: Encoding> Trits<T> {
+    pub fn new() -> Self {
+        Self { inner: T::with_capacity(0) }
+    }
+
+    pub fn with_capacity(n: usize) -> Self {
+        Self { inner: T::with_capacity(n) }
+    }
+
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    ...
+}
+
+impl<T: Encoding> Deref for Trits<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl From<Trits<T3B1>> for Trits<T1B1> {
+    ...
+}
+```
+
+# Drawbacks
+
+We define a general interface with different encoded types within to make
+consistency of public usage. But there are some method may needed for one
+encoded type but not others. Take hashing for example, types like `ptrits` would
+need some ways to do permutations and transformations, while types like 5T1B or
+9T2B are not required because they are mainly used for transfer.
+
+We can implement more methods but it still depends on what trait `Encoding` can
+provide. Or we give more freedom to have method like `into_inner` to manipulate
+the inner structure directly, but this might contradict to the purpose of this
+interface.
+
+# Rationale and alternatives
+
+- Just define type alias and some convert functions. This give way more freedom to the users.
+- Each encoded trit is a type of its own instead of interface like above.
+
+# Unresolved questions
+
+- Do we want to type usual sizes for more safety ?
+- In case of missing conversion, how are fallbacks to slower implementation handled ?
```

Comment:
IMHO this can be handled by the API user. He realizes a direct conversion is not supported, so he simply has to do 2 conversion steps (source_enc -> bridge_enc -> target_enc)

Reply:
I'm fine with that for lack of better alternative.